### ISSUE TITLE
[] account_edi: Why account_edi is autoinstall?

### ISSUE THREAD

**Issue by nhomar** [odoo/odoo#71190]
Impacted versions:

>14.0

Steps to reproduce:

Install only accounting, account_edi is installed in a country that do not require (at all) any electronic invoice (CA and EEUU for example).

Current behavior:

Installing a module that is not used, introducing errors (that make me lose time on things I will never use never)

Expected behavior:

Install as explicit modules as possible without noisy code on my server.

Video/Screenshot link (optional):

Support ticket number submitted via odoo.com/help (optional):

**Comment by moylop260** [comment]
`account_ponto` is a module only used in Belgium but How this module is even installed in my customer in LATAM?

Because it is auto-install:
 - https://github.com/odoo/enterprise/blob/8e14e2a401ff4aef6855fa6fff2cee569f8b1fdd/account_ponto/__manifest__.py#L28

If `account_online_sync` module is installed

And `account_online_sync` is auto-install too if only account is installed
 - https://github.com/odoo/enterprise/blob/8e14e2a401ff4aef6855fa6fff2cee569f8b1fdd/account_online_sync/__manifest__.py#L24

Similar case for base_vat_autocomplete module.
It is not even used in LATAM but we have issues related because auto-install
 - e.g. https://github.com/odoo/odoo/pull/24395

Another similar case `Single Euro Payments Area (SEPA)` only if you install the International Bank Account Number module `base_iban`
 - https://github.com/odoo/odoo/blob/3780fa2af5d5f6cac91e419bcab69a253db280bd/addons/account_qr_code_sepa/__manifest__.py#L18

@qdp-odoo @smetl @oco-odoo @csnauwaert
You are the authors
Maybe you can help us to understand it

@xmo-odoo 
Is there a way to bypass this useless auto-installed modules for our countries in the ORM?

**Comment by csnauwaert** [comment]
@moylop260 
I'll let one of my colleague answer for `account_edi` but I'll answer for the `account_online_sync` and `account_ponto` part.

First of all, one small detail, `account_ponto` is not only for Belgium but for European countries (ponto was mainly for belgium but they are increasing their coverage fast and they do a lot of european countries now).
Now to explain why this module is auto_install by default, the reason is pretty simple, historically we had only two different provider to fetch transactions from customers banking institutions directly, plaid and yodlee. Plaid was only US and yodlee worldwide. Both provider have their own module (namely `account_plaid` and `account_yodlee`) and they share a common module (`account_online_sync`) which contains method that are generic. In previous version, user had the choice to choose to install one or another module directly from the settings but this caused trouble as end users don't know that plaid is for US and yodlee for other countries. It was confusing and many issues reached us because they were problem with user not installing the module or user installing only `account_online_sync` without the proper provider module and no bank were showing and so on. So when we had to add another provider to the mix we decided that it would be easier to only have one settings which is "automatically synchronize with your banking institution" and that would install all providers. I do not see why you would not want that as the only thing that it does is adding "banking institutions" to the supported list displayed in Odoo when trying to synchronize with a bank. Choosing to install only the relevant provider (just ponto if we are in Europe for example) would not display the US bank and it is entirely possible that a company in Belgium has bank account in other countries. Also you should also think from a end user perspective, since we don't want to confuse the end user with multiple settings in accounting for the synchronization (what is plaid/ponto/yodlee? We don't want multiple checkbox and so on). The user friendly option should be a single checkbox that says "Automatic import" (as you can see in the attached screenshot) and that would install all providers so that when the end user is looking for a bank, the list is always the same for everyone.
![New - Odoo](https://user-images.githubusercontent.com/7338483/119466117-9cbd2780-bd44-11eb-830a-f48afc97e183.png)

Now, we could argue with the fact that this option is installed by default, this is a management choice since we want most of our users to know from the beginning that they can use the synchronization to connect with their bank.

Therefore, let me agree to disagree that this a problem that should be fixed. I can understand why you may think that `account_ponto` should not be installed in some countries but this is the way it is for a good reason.

Anyway since we are on that subject, you should know that we recently added a new provider (SaltEdge) and more providers are coming for other countries worldwide. Therefore we decided to rewrite completely the synchronization module and we backported it in all supported version (v12 to v14). This new module `account_online_synchronization` works differently than before. Previously we had the logic of each provider (yodlee, plaid, ponto) directly in a module available in the end user database, now we moved that logic to a proxy belonging to Odoo so that means that we can add/fix/change providers more easily without updating every user databases (and for your information we already had to do that before as some provider completely changed their API and this is a nightmare to update every customer database in that case).
So your question about ponto is deprecated, just install the `account_online_synchronization` which is the new synchronization module that every users should now use.
Starting from v15, `account_online_sync, account_plaid, account_yodlee and account_ponto` module will be deleted and only the `account_online_synchronization` will stay and all providers will be available for everyone by default.
From v12 to 14 we could not delete those module in stable, so the `account_online_synchronization` is just a module that is installing itself above those and also deactivate some method/views of the old synchronization.

Hope that this answer your question on why `account_ponto` is installed by default.

**Comment by bfr-o** [comment]
As for account_edi, the main reason is that facturx is enabled by default for all database regardless of the country. This is because historically this is the chosen format for import/export of invoice between odoo databases, and should be kept that way for now. Other edi modules are not auto_install. As for account_edi_ubl, I think your are right, we could remove the 'auto_install'.

**Comment by moylop260** [comment]
So, you have good reasons to auto-install these modules and it is fine (I think so)

What about if we have good reasons to non-install these modules?
  - I don't use European modules in non-European countries
  - There is a corner case raising errors when this module is installed
  - I have a custom/OCA module replacing the feature
  - There is performance issues extra computing methods/fields that I'm not using
  - There is incompatibility with other modules
  - ...

The problem with `auto_installing=True` is that there is not an easy way to let it uninstalled.

I mean,
Install `base_iban` module
Uninstall `account_qr_code_sepa` module
Update `base` module
You will have `account_qr_code_sepa` module installed again

IMHO -=+
Migrating version has a lot of corner cases that auto-installing modules raising errors is a lot of extra work.

But I understand it could be good
What about let us to decide early don't install them?

I tried to use the feature to exclude:
 - https://github.com/odoo/odoo/pull/16048
But it only raised a new error trying to install `account` because it is for another proposal


# OCA module `module_change_auto_install`

For record, OCA module `module_change_auto_install` to bypass auto-install and let us to decide:
 - https://github.com/OCA/server-tools/pull/2091
Thanks @legalsylvain

Note: if we have configured as `auto_install=False`
We even can decide to install it in the future (like other modules/apps)

Otherwise, you can not uninstall them in a peaceful way

**Comment by csnauwaert** [comment]
This is getting out of scope of the initial question which was why is account_edi installed.
This is because of facturx which is the standard of Import/Export for invoices between Odoo databases. This won't change in stable at the moment.

Now concerning the other point, I'm pretty sure that I know what the upper management will answer to the proposition of having an option to allow some auto_install module to be bypassed and not installed and it will probably be a big NO because other options exists.

What could probably be done:
- Modules like account_sepa or european modules should only be installed if we are in Europe (thanks to the post_init_hook in account.py). If it is not the case, there might be some changes that we can do for that. The exemple for account_ponto is already fixed in master (but again this module does nothing except printing more institutions to a list of supported bank).
- If there is a corner case raising error when the module is installed we should obviously try to fix it. In that case let us know of the problem and we'll do our best to fix it in stable or for the next version.
- custom OCA module used to replace the feature should be able to install above the existing module and overwrite the methods of the original module.
- Performance issues/fields I'm not using: we're trying our best to solve performance issues and we have people dedicated to that. Concerning fields that some people might not be using, they can be hidden from the view using custom modules.
- incompatibility with other module, I suppose you're speaking for incompatibility with OCA or custom modules. We're trying our best to make our module working together, in some case it is true that the code is not segmented enough and it is difficult to bypass completely a feature, we can hear that and try to refactor those part if there is a real need for that. Otherwise I'm pretty sure that there should be some way to make things works between multiple modules.

According to what I've heard, auto_install module should only be taken into account when doing a `-i` and not a `-u` (this should be verified as I haven't tested it).
If the reason to have this bypass for auto_install feature is just to remove iap or partner_autocomplete, it would probably not be accepted as those two modules are features that are key point of Odoo and are displayed on every presentation to sell the product. Pretty sure it won't change. We should already be happy that we managed to add most of the auto_install=true features in separate module, they could have been integrated directly within the main module :fearful: 
Allowing to bypass the auto_install feature could also cause problem during our own migration (or unforeseen problem/corner case) so in any case if that proposition was to be discussed it would need to be completely thing through.

I've created a feedback internally so that it could be discussed but don't expect much as I'm pretty sure that bypassing auto_install will probably not be accepted.

Anyway, since the original issue has been answered and that there was a reason for installing the account_edi module, I'm closing this.
